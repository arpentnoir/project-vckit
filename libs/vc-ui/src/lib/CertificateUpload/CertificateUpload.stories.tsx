import type { ComponentMeta, ComponentStory } from '@storybook/react';
import { CertificateUpload } from './CertificateUpload';

const Story: ComponentMeta<typeof CertificateUpload> = {
  component: CertificateUpload,
  title: 'CertificateUpload',
};
export default Story;

export const Primary: ComponentStory<typeof CertificateUpload> = () => (
  <CertificateUpload acceptMessage="Max 3MB" handleFiles={() => null} />
);

export const Error: ComponentStory<typeof CertificateUpload> = () => (
  <CertificateUpload
    errorMessage="Something went wrong"
    handleFiles={() => null}
  />
);
