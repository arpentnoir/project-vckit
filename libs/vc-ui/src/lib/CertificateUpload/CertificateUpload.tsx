import {
  Alert,
  Box,
  Input,
  InputLabel,
  Stack,
  SxProps,
  useTheme,
} from '@mui/material';
import React, {
  DragEvent,
  FormEvent,
  useEffect,
  useMemo,
  useRef,
  useState,
} from 'react';
import OpenBoxSVG from '../../assets/OpenBox.svg';
import { Text } from '../Text';

interface ICertificateUpload {
  handleFiles: (files: FileList) => void;
  errorMessage?: string;
  acceptMessage?: string;
  style?: SxProps;
}

export const CertificateUpload = ({
  handleFiles,
  errorMessage,
  acceptMessage,
  style,
}: ICertificateUpload) => {
  const [dragActive, setDragActive] = useState(false);
  const inputRef = useRef<HTMLInputElement>(null);
  const errorRef = useRef<HTMLParagraphElement>(null);

  const theme = useTheme();

  useEffect(() => {
    if (errorMessage && errorRef.current) {
      errorRef.current.focus();
    }
  }, [errorMessage]);

  const handleDrag = function (
    event: DragEvent<HTMLDivElement | HTMLFormElement>
  ) {
    event.preventDefault();
    event.stopPropagation();
    if (event.type === 'dragenter' || event.type === 'dragover') {
      setDragActive(true);
    } else if (event.type === 'dragleave') {
      setDragActive(false);
    }
  };

  const handleDrop = function (event: DragEvent<HTMLDivElement>) {
    event.preventDefault();
    event.stopPropagation();
    setDragActive(false);
    if (event?.dataTransfer?.files && event.dataTransfer.files[0]) {
      handleFiles(event.dataTransfer.files);
    }
  };

  const handleChange = function (
    event: FormEvent<HTMLInputElement | HTMLTextAreaElement>
  ) {
    event.preventDefault();
    const target = event?.target as HTMLInputElement;
    if (target.files && target.files[0]) {
      handleFiles(target.files);
    }
  };

  const onButtonClick = (event: any) => {
    event.preventDefault();
    if (event.key === ' ' || event.key === 'Enter') {
      inputRef?.current?.click();
    }
  };

  const textColor = useMemo(
    () => (errorMessage ? 'white' : 'black'),
    [errorMessage]
  );

  return (
    <React.Fragment>
      <Box
        onDragEnter={handleDrag}
        id="certificate-upload"
        sx={{
          height: '16rem',
          textAlign: 'center',
        }}
      >
        <Input
          inputRef={inputRef}
          type="file"
          id="input-file-upload"
          inputProps={{ 'data-testid': 'input-file-upload' }}
          onChange={handleChange}
          sx={{ display: 'none' }}
        />

        <InputLabel
          id="label-file-upload"
          htmlFor="input-file-upload"
          sx={{
            height: '100%',
            display: 'flex',
            alignItems: 'center',
            justifyContent: 'center',
            borderWidth: '2px',
            borderRadius: '1rem',
            borderStyle: 'dashed',
            borderColor: '#cbd5e1',
            backgroundColor: errorMessage
              ? theme.palette.error.main || '#AD1A1F'
              : 'white',
            // Override style
            ...style,
          }}
          aria-label="Upload file"
        >
          <Stack
            sx={{
              justifyContent: 'space-evenly',
              height: '100%',
            }}
          >
            <Stack
              sx={{
                justifyContent: 'space-evenly',
                alignItems: 'center',
                height: '100%',
                color: textColor,
              }}
            >
              <Box
                component="img"
                sx={{
                  maxHeight: 61,
                }}
                src={OpenBoxSVG as string}
              />
              <Text
                fontWeight="bold"
                color={textColor}
                variant="h6"
                tabIndex={-1}
              >
                Drop your document to verify
              </Text>
              {errorMessage && (
                <span ref={errorRef} aria-label={errorMessage} tabIndex={0} />
              )}

              <Box>
                <Text
                  sx={{
                    display: 'inline',
                  }}
                  variant="body1"
                  tabIndex={-1}
                >
                  Or &nbsp;
                </Text>
                <Box
                  role="button"
                  aria-controls="filename"
                  tabIndex={0}
                  aria-label={`Select a file. ${acceptMessage}`}
                  onKeyUp={onButtonClick}
                  sx={{
                    display: 'inline',
                    borderBottom: '1px solid',
                    ':hover': { cursor: 'pointer' },
                  }}
                >
                  Select a file
                </Box>
              </Box>

              {acceptMessage && (
                <Text variant="body1" tabIndex={-1}>
                  {acceptMessage}
                </Text>
              )}
              {errorMessage && (
                <Alert
                  severity="error"
                  id="vc-upload-error"
                  aria-hidden
                  sx={{ width: 'fit-content' }}
                >
                  {errorMessage}
                </Alert>
              )}
            </Stack>
          </Stack>
        </InputLabel>
        {dragActive && (
          <Box
            sx={{
              position: 'absolute',
              height: ' 100%',
              borderRadius: '1rem',
              top: '0px',
              right: '0px',
              bottom: '0px',
              left: '0px',
            }}
            id="drag-file-element"
            onDragEnter={handleDrag}
            onDragExit={handleDrag}
            onDragOver={handleDrag}
            onDrop={handleDrop}
          />
        )}
      </Box>
    </React.Fragment>
  );
};
