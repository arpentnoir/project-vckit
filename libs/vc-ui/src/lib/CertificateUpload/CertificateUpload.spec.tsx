import { fireEvent, render, waitFor } from '@testing-library/react';
import { CertificateUpload } from './CertificateUpload';

const handleFiles = jest.fn();
const ACCEPT_MESSAGE = 'Max 3MB';
const ERROR_MESSAGE = 'Something went wrong';

describe('CertificateUpload', () => {
  it('should render correctly', () => {
    const { baseElement } = render(
      <CertificateUpload
        handleFiles={handleFiles}
        acceptMessage={ACCEPT_MESSAGE}
        errorMessage=""
      />
    );

    expect(baseElement).toMatchSnapshot();
  });

  it('should display error message if present', () => {
    let baseElement = render(
      <CertificateUpload
        handleFiles={handleFiles}
        acceptMessage={ACCEPT_MESSAGE}
        errorMessage=""
      />
    );

    expect(baseElement.queryByText('Something went wrong')).toBeFalsy();

    baseElement = render(
      <CertificateUpload
        handleFiles={handleFiles}
        acceptMessage={ACCEPT_MESSAGE}
        errorMessage={ERROR_MESSAGE}
      />
    );

    expect(baseElement.queryByText('Something went wrong')).toBeTruthy();
  });

  it('should call handleFiles on upload', async () => {
    const baseElement = render(
      <CertificateUpload
        handleFiles={handleFiles}
        acceptMessage={ACCEPT_MESSAGE}
        errorMessage=""
      />
    );

    const uploadElement = baseElement.getByTestId('input-file-upload');

    await waitFor(() =>
      fireEvent.change(uploadElement, {
        target: { files: ['{"test": "hello"}'] },
      })
    );

    expect(handleFiles).toHaveBeenCalledWith(['{"test": "hello"}']);
  });
});
