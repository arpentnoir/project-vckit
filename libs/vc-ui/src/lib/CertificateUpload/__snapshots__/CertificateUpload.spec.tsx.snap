// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CertificateUpload should render correctly 1`] = `
<body>
  <div>
    <div
      class="MuiBox-root css-1qj0fvz"
      id="certificate-upload"
    >
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary css-1g6u8k6-MuiInputBase-root-MuiInput-root"
      >
        <input
          class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
          data-testid="input-file-upload"
          id="input-file-upload"
          type="file"
          value=""
        />
      </div>
      <label
        aria-label="Upload file"
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-animated MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-animated css-z5ab98-MuiFormLabel-root-MuiInputLabel-root"
        for="input-file-upload"
        id="label-file-upload"
      >
        <div
          class="css-hnuxw1-MuiStack-root"
        >
          <div
            class="css-1q46w03-MuiStack-root"
          >
            <img
              class="MuiBox-root css-eu33wj"
              src="OpenBox.svg"
            />
            <h6
              class="MuiTypography-root MuiTypography-h6 css-n8wto6-MuiTypography-root"
              tabindex="-1"
            >
              Drop your document to verify
            </h6>
            <div
              class="MuiBox-root css-0"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-1sx1eu0-MuiTypography-root"
                tabindex="-1"
              >
                Or Â 
              </p>
              <div
                aria-controls="filename"
                aria-label="Select a file. Max 3MB"
                class="MuiBox-root css-13imy7l"
                role="button"
                tabindex="0"
              >
                Select a file
              </div>
            </div>
            <p
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
              tabindex="-1"
            >
              Max 3MB
            </p>
          </div>
        </div>
      </label>
    </div>
  </div>
</body>
`;
